src/stream/forward.rs

```diff
@@ -0,0 +1,90 @@
+use {Poll, Async, Future, AsyncSink};
+use stream::{Stream, Fuse};
+use sink::Sink;
+
+/// Future for the `Stream::forward` combinator, which sends a stream of values
+/// to a sink and then waits until the sink has fully flushed those values.
+#[must_use = "futures do nothing unless polled"]
+pub struct Forward<T: Stream, U> {
+    sink: Option<U>,
+    stream: Option<Fuse<T>>,
+    buffered: Option<T::Item>,
+}
+
+
+pub fn new<T, U>(stream: T, sink: U) -> Forward<T, U>
+    where U: Sink<SinkItem=T::Item>,
+          T: Stream,
+          T::Error: From<U::SinkError>,
+{
+    Forward {
+        sink: Some(sink),
+        stream: Some(stream.fuse()),
+        buffered: None,
+    }
+}
+
+impl<T, U> Forward<T, U>
+    where U: Sink<SinkItem=T::Item>,
+          T: Stream,
+          T::Error: From<U::SinkError>,
+{
+    fn sink_mut(&mut self) -> &mut U {
+        self.sink.as_mut().take()
+            .expect("Attempted to poll Forward after completion")
+    }
+
+    fn stream_mut(&mut self) -> &mut Fuse<T> {
+        self.stream.as_mut().take()
+            .expect("Attempted to poll Forward after completion")
+    }
+
+    fn take_result(&mut self) -> (T, U) {
+        let sink = self.sink.take()
+            .expect("Attempted to poll Forward after completion");
+        let fuse = self.stream.take()
+            .expect("Attempted to poll Forward after completion");
+        return (fuse.into_inner(), sink)
+    }
+
+    fn try_start_send(&mut self, item: T::Item) -> Poll<(), U::SinkError> {
+        debug_assert!(self.buffered.is_none());
+        if let AsyncSink::NotReady(item) = try!(self.sink_mut().start_send(item)) {
+            self.buffered = Some(item);
+            return Ok(Async::NotReady)
+        }
+        Ok(Async::Ready(()))
+    }
+}
+
+impl<T, U> Future for Forward<T, U>
+    where U: Sink<SinkItem=T::Item>,
+          T: Stream,
+          T::Error: From<U::SinkError>,
+{
+    type Item = (T, U);
+    type Error = T::Error;
+
+    fn poll(&mut self) -> Poll<(T, U), T::Error> {
+        // If we've got an item buffered already, we need to write it to the
+        // sink before we can do anything else
+        if let Some(item) = self.buffered.take() {
+            try_ready!(self.try_start_send(item))
```

Comment:
Well, this is a good question. It's not clear for me that `start_send` never does any I/O. Because otherwise every failure of `start_send` is basically a `start_send -> poll_complete -> start_send`.

From what I can see:
1. `tokio_core::Framed` calls poll_complete under pressure.
2. `sink::Send` doesn't call poll_complete on `AsyncSink::NotReady`
3. `SendAll` has same behavior (because I've copied from there)
4. `sink::Buffer` always tries to flush on start_send (which is fine), and does `poll_complete` on `NotReady` [*]

[*] So only a buffer expects this semantics. And yes if we rely on assumption of `start_send` doing I/O under pressure, we have a bug in Buffer (which ironically is fixed by removing `poll_complete` in `Buffer::start_send`)